generator/src/main/java/com/graphicsfuzz/generator/mutateapi/BooleanValue.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 The GraphicsFuzz Project Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.graphicsfuzz.generator.mutateapi;
+
+import com.graphicsfuzz.common.ast.expr.BoolConstantExpr;
+import com.graphicsfuzz.common.ast.expr.Expr;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.Type;
+import com.graphicsfuzz.generator.semanticschanging.LiteralFuzzer;
+import java.util.Optional;
+
+public class BooleanValue implements Value {
+
+  private final Optional<Boolean> value;
+  private final BasicType type;
+
+  public BooleanValue(Optional<Boolean> value) {
+    this.value = value;
+    this.type = BasicType.BOOL;
+  }
+
+  @Override
+  public boolean equals(Object that) {
+    return that instanceof BooleanValue && equals((BooleanValue) that);
+  }
+
+  public boolean equals(BooleanValue that) {
+    if (this == that) {
+      return true;
+    }
+
+    if (!this.valueIsKnown() && !that.valueIsKnown()) {
+      return true;
+    }
+
+    return this.getValue() == that.getValue();
+  }
+
```

Comment:
Don't forget to override hashCode() - you always need to do this when you override equals.

Reply:
I implemented hashCode with the idea of using prime numbers here. I am not quite sure how you do it normally. Please let me know what you think.

Reply:
This is overkill.

First: there's no point doing "getType().hashCode()" since "getType()" always returns the same thing.

Second: Optional already has its own "hashCode()" method.

So you can just do:

@Override
public int hashCode() {
    return value.hashCode();
}